bench: refactor random number generation in `stats/base/dists/bernoulli`

anandkaranubc opened the pull request:
Resolves none.

## Description

> What is the purpose of this pull request?

This pull request:

- Refactors random number generation in JS benchmarks for `stats/base/dists/bernoulli`.
- Replaces `randu()` with `uniform()` for cleaner and more consistent code.

## Related Issues

> Does this pull request have any related issues?

This pull request:

- resolves none

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

anandkaranubc committed:
refactor: random number generation in JS benchmarks for stats/base/dists/bernoulli

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: passed
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: passed
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

anandkaranubc commented:
Hi @kgryte,

I wanted to confirm a couple of things about these refactorings:

1. Are we replacing **all `randu()` expressions** with `uniform()` for consistency, or only those where the **minimum cannot be zero** (e.g., `randu() * 1.0 + EPS`)?

2. For expressions like `round( randu() )`, would it be better to use `discreteUniform` instead?

Thanks for the help!

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/bernoulli/mean/benchmark/benchmark.js:
```diff
@@ -40,7 +40,7 @@ bench( pkg, function benchmark( b ) {
 	p = new Float64Array( len );
 	for ( i = 0; i < len; i++ ) {
-		p[ i ] = ( randu()*1.0 ) + EPS;
+		p[ i ] = uniform( EPS, EPS + 1.0 );
```
```suggestion
		p[ i ] = uniform( EPS, 1.0 );
```

`p` is a success probability, so you never want it to exceed `1`. The previous worked, as `randu()` is always `< 1`.

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/bernoulli/mean/benchmark/benchmark.native.js:
```diff
@@ -49,7 +49,7 @@ bench( pkg+'::native', opts, function benchmark( b ) {
 	p = new Float64Array( len );
 	for ( i = 0; i < len; i++ ) {
-		p[ i ] = ( randu()*1.0 ) + EPS;
+		p[ i ] = uniform( EPS, EPS + 1.0 );
```
```suggestion
		p[ i ] = uniform( EPS, 1.0 );
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM

kgryte commented:
> Are we replacing all randu() expressions with uniform() for consistency, or only those where the minimum cannot be zero (e.g., randu() * 1.0 + EPS)?

All would be nice, not just those involving `EPS`.

> For expressions like round( randu() ), would it be better to use discreteUniform instead?

Yes, discrete-uniform would be preferred.

anandkaranubc commented:
> > Are we replacing all randu() expressions with uniform() for consistency, or only those where the minimum cannot be zero (e.g., randu() * 1.0 + EPS)?
> 
> All would be nice, not just those involving `EPS`.
> 
> > For expressions like round( randu() ), would it be better to use discreteUniform instead?
> 
> Yes, discrete-uniform would be preferred.

I meant even the ones like `randu() * 20.0` or leave them like that?

kgryte commented:
> I meant even the ones like randu() * 20.0 or leave them like that?

Yes, even those.

anandkaranubc commented:
> > I meant even the ones like randu() * 20.0 or leave them like that?
> 
> Yes, even those.

Open to your feedback, @kgryte. Thanks!

anandkaranubc committed:
refactor: improve random number generation in JS benchmarks for stats/base/dists/bernoulli

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

anandkaranubc committed:
Merge branch 'bernoulli-benchmarks' of https://github.com/anandkaranubc/stdlib into bernoulli-benchmarks

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/bernoulli/cdf/benchmark/benchmark.js:
```diff
@@ -37,8 +37,8 @@ bench( pkg, function benchmark( b ) {
 	for ( i = 0; i < b.iterations; i++ ) {
-		x = ( randu()*3.0 ) - 1.0;
-		p = randu();
+		x = uniform( -1.0, 2.0 );
```
@anandkaranubc Would you mind moving this RV generation outside the loop? Otherwise, the benchmark is just as likely to be measuring RV generation as it is the performance of computing the `cdf`. That applies here and in other benchmarks.

anandkaranubc replied:
Should be ready for review! Thanks!

kgryte commented on lib/node_modules/@stdlib/stats/base/dists/bernoulli/mode/benchmark/benchmark.js:
```diff
@@ -39,7 +39,7 @@ bench( pkg, function benchmark( b ) {
 	p = new Float64Array( len );
 	for ( i = 0; i < len; i++ ) {
-		p[ i ] = randu();
+		p[ i ] = uniform( 0.0, 1.0 );
```
@anandkaranubc This is the desired approach for RV generation. Notice that we create an array of values outside the benchmark loop.

anandkaranubc replied:
Working on it! I was initially thinking about doing it right after this in the next PR. Thanks for the go-ahead to include it in this PR only!

kgryte replied:
Thanks for doing this Karan!

anandkaranubc committed:
bench: move random number generation outside the benchmarking loops

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte reviewed (approved):
LGTM. Thanks, @anandkaranubc!
